src/coreclr/jit/importer.cpp

```diff
@@ -9031,7 +9038,36 @@ void Compiler::impImportBlockCode(BasicBlock* block)
                         //        replacing the pattern with a call to a thunk, which contains roughly the same code.
 
                         const BYTE* nextOpcode = codeAddr + sizeof(mdToken);
-                        if (compIsAsync2() && (nextOpcode + sizeof(mdToken) < codeEndp) &&
+                        // There must be enough space after ldc for {call + tk + call + tk}
+                        if (nextOpcode + 2 * (1 + sizeof(mdToken)) < codeEndp)
+                        {
+                            uint8_t nextOp = getU1LittleEndian(nextOpcode);
+                            uint8_t nextNextOp = getU1LittleEndian(nextOpcode + 1);
+                            if ((nextOp != CEE_LDC_I4_0 && nextOp != CEE_LDC_I4_1) ||
+                                (nextNextOp != CEE_CALL && nextNextOp != CEE_CALLVIRT))
+                            {
+                                goto checkForAwait;
+                            }
+
+                            // check if the token after {ldc, call[virt]} is ConfigAwait
+                            CORINFO_RESOLVED_TOKEN nextCallTok;
+                            impResolveToken(nextOpcode + 2, &nextCallTok, CORINFO_TOKENKIND_Method);
+
+                            if (!eeIsIntrinsic(nextCallTok.hMethod) ||
+                                lookupNamedIntrinsic(nextCallTok.hMethod) !=
+                                    NI_System_Threading_Tasks_Task_ConfigureAwait)
+                            {
+                                goto checkForAwait;
+                            }
+
+                            configVal = nextOp == CEE_LDC_I4_0 ? 0 : 1;
+                            // skip {ldc; call; <ConfigureAwait>}
+                            nextOpcode += 1 + 1 + sizeof(mdToken);
+                        }
+
+                    checkForAwait:
```

Comment:
This logic is getting complicated enough that I think we should extract it all to a function.

Reply:
Moved the Await pattern-match out as a separate method.

Reply:
Are there more concerns or ok to merge?